### PR TITLE
Feature/sdf2sdf

### PR DESCRIPTION
**Pull request by fei-shan**
Review. sdf2sdf optimizer is tested in python code.

### CODE REVIEW

File: src/math/transformation.cpp [hunk 1]
```diff
@@ -0,0 +1,45 @@
...
+        q = eig::Quaternionf(cos(theta/2),
+                             sin(theta/2) * rotation(0),
+                             sin(theta/2) * rotation(1),
+                             sin(theta/2) * rotation(2));
+
+        eig::Matrix3f rotation_matrix = q.toRotationMatrix();
+
+        eig::Matrix4f twist_matrix_homo;
+
+        twist_matrix_homo << rotation_matrix(0, 0), rotation_matrix(0, 1), rotation_matrix(0, 2), translation(0),
```
Comment (Algomorph) [inline src/math/transformation.cpp]: please refactor/rename this variable to "twist_matrix_homography" or "twist_matrix_transformation". "homo" has connotations that are unwanted here.

File: src/rigid_optimization/sdf_2_sdf_optimizer2d.cpp [hunk 2]
```diff
@@ -0,0 +1,151 @@
...
+            voxel_size(voxel_size),
+            narrow_band_width_voxels(narrow_band_width_voxels)
+{
+};
+
+eig::Vector3f Sdf2SdfOptimizer2d::optimize(int image_y_coordinate,
+        const eig::Matrix<unsigned short, eig::Dynamic, eig::Dynamic>& canonical_depth_image,
+        const eig::Matrix<unsigned short, eig::Dynamic, eig::Dynamic>& live_depth_image,
+        const Sdf2SdfOptimizer2d::TSDFGenerationParameters tsdf_generation_parameters,
+        float eta = 0.01f) {
```
Comment (Algomorph) [inline src/rigid_optimization/sdf_2_sdf_optimizer2d.cpp]: Default arguments should be defined in the header and only in the header, not the .cpp file.

File: src/rigid_optimization/sdf_2_sdf_optimizer2d.cpp [hunk 3]
```diff
@@ -0,0 +1,151 @@
...
+        for (int j=0; j < canonical_weight.cols(); ++j){
+            canonical_weight(i, j) = (canonical_field(i, j) <= -eta) ? 0.0f : 1.0f;
+        }
+    }
+
+    eig::Vector3f twist = eig::Vector3f(0.0f, 0.0f, 0.0f);
+
+    for (int iteration_count=0; iteration_count<maximum_iteration_count; ++iteration_count) {
+        eig::Matrix3f matrix_A = eig::Matrix3f::Zero();
+        eig::Vector3f vector_b = eig::Vector3f::Zero();
```
Comment (Algomorph) [inline src/rigid_optimization/sdf_2_sdf_optimizer2d.cpp]: Are there any names that are more descriptive than "matrix_A",  "vector_b", and "twist_star"?

File: src/rigid_optimization/sdf_2_sdf_optimizer2d.cpp [hunk 4]
```diff
@@ -0,0 +1,151 @@
...
+        eig::Matrix<eig::Vector3f, eig::Dynamic, eig::Dynamic> live_gradient(canonical_field.rows(),
+                                                                             canonical_field.cols());
+        ropt::gradient_wrt_twist(live_field,
+                                 twist,
+                                 tsdf_generation_parameters.array_offset,
+                                 tsdf_generation_parameters.voxel_size,
+                                 live_gradient);
+
+        for (int i=0; i < live_field.rows(); ++i) {
+            for (int j=0; j < live_field.cols(); ++j) {
```
Comment (Algomorph) [inline src/rigid_optimization/sdf_2_sdf_optimizer2d.cpp]: Eigen matrices are column-major, so, for data locality, it makes better sense to traverse in a column-major order, i.e. rows would be the innter loop.

File: src/rigid_optimization/sdf_gradient_wrt_transformation2d.cpp [hunk 5]
```diff
@@ -0,0 +1,54 @@
...
+                            eig::Matrix<eig::Vector3f, eig::Dynamic, eig::Dynamic>& gradient_field) {
+
+        math::MatrixXv2f gradient_first_term;
+        math::gradient(gradient_first_term, live_field);
+        eig::Matrix3f twist_matrix_homo_inv2d = math::transformation_vector_to_matrix2d(-twist);
+
+        float x_voxel, z_voxel, w_voxel = 1;
+
+        for (int y_field=0; y_field<live_field.rows(); ++y_field) {
+            for (int x_field=0; x_field<live_field.cols(); ++x_field) {
```
Comment (Algomorph) [inline src/rigid_optimization/sdf_gradient_wrt_transformation2d.cpp]: Same comment as above; it makes more sense to do column-major traversal with Eigen

File: src/rigid_optimization/sdf_gradient_wrt_transformation2d.cpp [hunk 6]
```diff
@@ -0,0 +1,54 @@
...
+//                gradient_field(y_field, x_field) = eig::Vector3f(0.0f, 0.0f, 0.0f);
```
Comment (Algomorph) [inline src/rigid_optimization/sdf_gradient_wrt_transformation2d.cpp]: Remove dead code

File: src/rigid_optimization/sdf_gradient_wrt_transformation2d.cpp [hunk 7]
```diff
@@ -0,0 +1,54 @@
...
+
+                x_voxel = (x_field + array_offset[0]) * voxel_size; // x coordinate
+                z_voxel = (y_field + array_offset[2]) * voxel_size; // z coordinate
+
+                eig::Vector3f point(x_voxel, z_voxel, w_voxel);
+                eig::Vector3f trans_point = twist_matrix_homo_inv2d * point;
```
Comment (Algomorph) [inline src/rigid_optimization/sdf_gradient_wrt_transformation2d.cpp]: see prev. comment about the "homo"

File: src/tsdf/tsdf_2d.cpp [hunk 8]
```diff
@@ -0,0 +1,84 @@
...
+
+    int matrix_size = static_cast<int>(field.size());
+
+#pragma omp parallel for
+    for (int i_element = 0; i_element < matrix_size; i_element++) {
+        // Any MatrixXf in Eigen is column-major
+        // i_element = x * column_count + y
+        div_t division_result = std::div(i_element, field_size);
+        int x_field = division_result.quot;
+        int y_field = division_result.rem;
```
Comment (Algomorph) [inline src/tsdf/tsdf_2d.cpp]: please revise to not use std::div operation, use "%" and "/" operators instead. Compiler will auto-optimize those to a single operation.

File: tests/test_tsdf.cpp [hunk 9]
```diff
@@ -282,3 +283,63 @@ BOOST_AUTO_TEST_CASE(test_EWA_3D_generation02) {
...
+			0.001f, //depth unit ratio
+			camera_intrinsic_matrix,
+			eig::Matrix4f::Identity(), //camera pose
+			offset,
+			field_size, //field size
+			0.004f, //voxel size
+			20, // narrow band width
+			-999.f
+			);
+//	std::cout << field.format(eig::IOFormat(eig::FullPrecision, 0, "f, ", "f,\n")) << std::endl;
```
Comment (Algomorph) [inline tests/test_tsdf.cpp]: remove dead code

File: tests/test_tsdf.cpp [hunk 10]
```diff
@@ -282,3 +283,63 @@ BOOST_AUTO_TEST_CASE(test_EWA_3D_generation02) {
...
+			0.001f, //depth unit ratio
+			camera_intrinsic_matrix,
+			camera_extrinsic_matrix, //camera pose
+			offset,
+			field_size, //field size
+			0.004f, //voxel size
+			20, // narrow band width
+			-999.f
+	);
+//	std::cout << field.format(eig::IOFormat(eig::FullPrecision, 0, "f, ", "f,\n")) << std::endl;
```
Comment (Algomorph) [inline tests/test_tsdf.cpp]: remove dead code

**Review by Algomorph — CHANGES_REQUESTED** [review]
Code looks good overall, I request a few minor changes. Excited to see how it works!

File: src/math/transformation.cpp (hunk 1 above)
Comment (fei-shan) [reply src/math/transformation.cpp]: fixed.

File: src/rigid_optimization/sdf_2_sdf_optimizer2d.cpp (hunk 2 above)
Comment (fei-shan) [reply src/rigid_optimization/sdf_2_sdf_optimizer2d.cpp]: fixed

File: src/rigid_optimization/sdf_2_sdf_optimizer2d.cpp (hunk 3 above)
Comment (fei-shan) [reply src/rigid_optimization/sdf_2_sdf_optimizer2d.cpp]: The A and b are from the paper, they are used to calculate the optimal twist, I change the twist_star to optimal_twist.

File: src/rigid_optimization/sdf_2_sdf_optimizer2d.cpp (hunk 4 above)
Comment (fei-shan) [reply src/rigid_optimization/sdf_2_sdf_optimizer2d.cpp]: fixed

File: src/rigid_optimization/sdf_gradient_wrt_transformation2d.cpp (hunk 5 above)
Comment (fei-shan) [reply src/rigid_optimization/sdf_gradient_wrt_transformation2d.cpp]: fixed

File: src/rigid_optimization/sdf_gradient_wrt_transformation2d.cpp (hunk 6 above)
Comment (fei-shan) [reply src/rigid_optimization/sdf_gradient_wrt_transformation2d.cpp]: fixed

File: src/rigid_optimization/sdf_gradient_wrt_transformation2d.cpp (hunk 7 above)
Comment (fei-shan) [reply src/rigid_optimization/sdf_gradient_wrt_transformation2d.cpp]: fixed

File: src/tsdf/tsdf_2d.cpp (hunk 8 above)
Comment (fei-shan) [reply src/tsdf/tsdf_2d.cpp]: fixed

File: tests/test_tsdf.cpp (hunk 9 above)
Comment (fei-shan) [reply tests/test_tsdf.cpp]: fixed

File: tests/test_tsdf.cpp (hunk 10 above)
Comment (fei-shan) [reply tests/test_tsdf.cpp]: fixed

**Comment by fei-shan** [comment]
I push a new commit.

**Review by Algomorph — APPROVED** [review]

**Comment by Algomorph** [comment]
Please rebase -i again on master, resolve merge conflicts/any test problems, and push.

**Comment by fei-shan** [comment]
Rebase on master

**Pull request merged**